testing/ui-tests/src/utils/metadata_test_runner.rs

```diff
@@ -2,72 +2,179 @@
 // This file is dual-licensed as Apache-2.0 or GPL-3.0.
 // see LICENSE for license details.
 
-use codec::Encode;
-use frame_metadata::RuntimeMetadataPrefixed;
+use codec::{Decode, Encode};
+use std::io::Read;
+use subxt_metadata::Metadata;
 
 static TEST_DIR_PREFIX: &str = "subxt_generated_ui_tests_";
+static METADATA_FILE: &str = "../../artifacts/polkadot_metadata_full.scale";
 
 #[derive(Default)]
 pub struct MetadataTestRunner {
     index: usize,
 }
 
 impl MetadataTestRunner {
-    pub fn path_to_ui_test_for_metadata(
-        &mut self,
-        name: impl AsRef<str>,
-        metadata: RuntimeMetadataPrefixed,
-    ) -> String {
-        let test_name = name.as_ref();
-
-        // increment test index to avoid overlaps.
+    /// Loads metadata that we can use in our tests. Panics if
+    /// there is some issue decoding the metadata.
+    pub fn load_metadata() -> Metadata {
+        let mut file =
+            std::fs::File::open(METADATA_FILE).expect("Cannot open metadata.scale artifact");
+
+        let mut bytes = Vec::new();
+        file.read_to_end(&mut bytes)
+            .expect("Failed to read metadata.scale file");
+
+        Metadata::decode(&mut &*bytes).expect("Cannot decode metadata bytes")
+    }
+
+    /// Create a new test case.
+    pub fn new_test_case(&mut self) -> MetadataTestRunnerCaseBuilder {
         let index = self.index;
+        // increment index so that each test case gets its own folder path.
         self.index += 1;
 
+        MetadataTestRunnerCaseBuilder::new(index)
+    }
+}
+
+// `trybuild` runs all tests once it's dropped. So, we defer all cleanup until we
+// are dropped too, to make sure that cleanup happens after tests are ran.
+impl Drop for MetadataTestRunner {
+    fn drop(&mut self) {
+        for i in 0..self.index {
+            let mut tmp_dir = std::env::temp_dir();
+            tmp_dir.push(format!("{TEST_DIR_PREFIX}{i}"));
+            std::fs::remove_dir_all(tmp_dir).expect("cannot cleanup temp files");
+        }
+    }
+}
+
+/// Build a single test case.
+pub struct MetadataTestRunnerCaseBuilder {
+    index: usize,
+    name: String,
+    validation_metadata: Option<Metadata>,
+    should_be_valid: bool,
+}
+
+impl MetadataTestRunnerCaseBuilder {
+    fn new(index: usize) -> Self {
+        MetadataTestRunnerCaseBuilder {
+            index,
+            name: format!("Test {index}"),
+            validation_metadata: None,
+            should_be_valid: true,
+        }
+    }
+
+    /// Set the test name.
+    pub fn name(mut self, name: impl AsRef<str>) -> Self {
+        self.name = name.as_ref().to_owned();
+        self
+    }
+
+    /// Set metadata to be validated against the generated code.
+    /// By default, we'll validate the same metadata used to generate the code.
+    pub fn validation_metadata(mut self, md: impl Into<Metadata>) -> Self {
```

Comment:
I think `set_metadata` is better name for this method...

Reply:
The `build()` call is where you pass the actual metadata that the macro will use; this `validation_metadata` is just the optional metadata you can use to test the `is_codegen_valid_for` function against :)